### ISSUE TITLE
[Feature request] Option for qTox to work as TCP-relay

### ISSUE THREAD

**Issue by zetok** [qTox/qTox#502]
Way to enable working as TCP-relay + way for user to set ports for TCP-relay function

**Comment by dubslow** [comment]
What does this mean, exactly?

Every client is a node in the network...

**Comment by zetok** [comment]
Yes, every client is a node in DHT, but apparently TCP relay function has to be explicitly enabled: https://github.com/irungentoo/toxcore/blob/6114bd7f3ef1fda71e45e81f259074cf4f8e58eb/other/bootstrap_daemon/tox-bootstrapd.c#L67 ← that and further in that file

**Comment by dubslow** [comment]
You misunderstand. That's only for the boostrap daemon -- TCP is enabled in Messenger.c, which is what clients use.

The main() in that file is analogous to, but different from, new_Messenger() (which is the same as tox_new() as used in clients).

**Comment by zetok** [comment]
@dubslow I haven't asked about "enabling" TCP.

I've asked to add to qTox TCP relay functionality, which according to @stqism isn't there.

**Comment by tux3** [comment]
I'm not aware of any way to enable TCP relay functionality through the toxcore API.

Did @stqism mention a way to do this ?

**Comment by stqism** [comment]
@tux3 It isn't part of the API, you have to import another header from toxcore to enable it. 

Whatever you do, DO NOT ADD IT. 

People will enable it on machines with dynamic IPs and or slow lines and potentially hurt the network. 

It really shouldn't be used unless you plan on keeping your client on for months at a time 24/7 without crashing and don't mind a massive increase in bandwidth.

**Comment by tux3** [comment]
I understand, thank you.
Tor has pretty much the same problems.

**Comment by zetok** [comment]
```
[15:45:34] <zetok> irungentoo_: is support for running TCP-server in clients a bad idea?
[15:46:30] <irungentoo_> no but it has to be done correctly
[15:46:44] <zetok> "correctly", as in coded properly?
[15:47:25] <zetok> I'm asking because of: https://github.com/tux3/qTox/issues/502#issuecomment-60273557
[15:52:08] <irungentoo_> it's mainly to make sure you don't get 1k+ connections when you enable it
[15:52:57] <zetok> big red warning with a lot of exclamation marks should do the job, right?
[15:53:07] <irungentoo_> somehow make the tcp server work only for your friends to connect to their friends or something
[15:53:31] <irungentoo_> that would be the best solution
```

Please disregard what @stqism said.

TCP-relay function in qTox would be really nice.

**Comment by tux3** [comment]
@zetok 

> Please disregard what @stqism said.

I'm not sure that's a good idea. Why should I do that ?

> <irungentoo_>somehow make the tcp server work only for your friends to connect to their friends or something

Is that possible right now ? Because I'd rather not start seeing tons of relays go up and down every 5 minutes on home connections. 
I'm not sure how that would be a good thing. 

Why not run the bootstrap daemon if you have a stable and fast connection ?
What's the use case ?

**Comment by zetok** [comment]
Use case is for users that are not technical enough to setup some strange bootstrap node, but they do have hardware and will to help Tox network.
(Yes, it is a lot easier to mark some checkbox in settings rather than getting a *nix server to run bootstrap node; not to mentions that ports don't even have to be opened manually in future, provided that PR for UPnP support will some day be merged -.- )

What @irungentoo said, is that problem for users running TCP relay, since they may get a lot of traffic, which could possibly affect their UX.

Also, since it's in no way a default setting, and users really should be advised against enabling it unless they do know what are they doing (big red warning), I don't really see a problem..

**Comment by tux3** [comment]
> Use case is for users that are not technical enough to setup some strange bootstrap node, but they do have hardware and will to help Tox network.

What I'm worried about is that users "not technical enough to setup some strange bootstrap node" might not be technical enough to understand whether their running a TCP relay on their personal computer will help the network or harm it.

> What @irungentoo said, is that problem for users running TCP relay, since they may get a lot of traffic, which could possibly affect their UX.

And users using the TCP relays may get a terrible UX too, instead of stable relays running on real servers maintained by 'technical enough' people :)

I'm not saying that we should make starting a TCP relay as hard as possible, far from it. 
But I'm not convinced that running a relay from a home computer will help the Tox network. And if you do have a server, why would go through the trouble of running a Qt client and changing the settings instead of just starting the bootstrap daemon ?

> Also, since it's in no way a default setting, and users really should be advised against enabling it unless they do know what are they doing (big red warning), I don't really see a problem..

That would be a solution, but unless you plan to run a Qt tox client on a server, that still implies having a relay on a home computer behind a home connection.

I have a tendency to write wall of text todays. 
TL;DR: I'm not sure we _want_ random people starting relays on their pc. Do we ? Why ?

**Comment by stqism** [comment]
@tux3 I'm speaking to @irungentoo in real life and he explained that even if you added tcp relay support, no one would connect to you in the first place. 

Also, he's planning on adding a tcp relay for friends feature that could be used in clients to help them Tox, but this will happen later.

**Comment by zetok** [comment]
Ok, basically I have real doubt about correctness of @stqism 's claim. Why is that?
Lets think for a moment. If Tox network could be easily disrupted by TCP relays going up and down often, what could possibly stop some entity from doing exactly that?

From what I remember, TCP relay system was supposed to be implemented in a way what would mitigate risk of disruption of "service". As in, not to use single TCP relay, to ensure that Tox network will still be available even in scenario where given TCP relay would go down.

If it's not done in reliable way, then Tox **will** inevitably go down, whenever someone who doesn't like Tox will feel like taking Tox down.

> That would be a solution, but unless you plan to run a Qt tox client on a server, that still implies having a relay on a home computer behind a home connection.

And what could be possibly wrong with that? It's no longer 90s, where there was lighting internet speed that used to buffer, and there are plenty of really high performance hardware in many households. Seriously, difference between a "server" in a cloud, where it has 1 core and 256 MiB of RAM, and typical new AMD PC box with 8 cores @ at least 3.5 GHz + SSD drive + 32 GB of memory.. Add to it botnet fiber+ and… I myself have a really hard time judging which one makes a better server…

But above doesn't even really matter.
Currently public Tox bootstraps (which have TCP-server enabled) generate ~250kbps of traffic.. That's really **a lot**, considering that they're public, and like every client uses them to bootstrap. And that CPU usage! I'm sure it goes into cosmic values…

I'm not sure about your internet connection, but I myself could have that on my home network without problems. And my internet connection isn't even that good..

**Comment by tux3** [comment]
@zetok 

> If Tox network could be easily disrupted by TCP relays going up and down often, what could possibly stop some entity from doing exactly that?

Nothing, I'd assume. Just like you could probably spam the DHT if you wanted to.

> If it's not done in reliable way, then Tox will inevitably go down, whenever someone who doesn't like Tox will feel like taking Tox down.

I agree, but that's not reason enough to add that option.

> > That would be a solution, but unless you plan to run a Qt tox client on a server, that still implies having a relay on a home computer behind a home connection.
> 
> And what could be possibly wrong with that? 

People might decide to use their home connection do download 350 TB of cat pictures or maybe shutdown their computers so it doesn't get too hot with qTox running. A server is supposed to be more or less consistently available.

> Currently public Tox bootstraps (which have TCP-server enabled) generate ~250kbps of traffic.. That's really a lot, considering that they're public, and like every client uses them to bootstrap. And that CPU usage! I'm sure it goes into cosmic values…

Yes, the network is still small. Hopefully, it'll get bigger.

> I'm not sure about your internet connection, but I myself cold have that on my home network without problems. And my internet connection isn't even that good..

But can I trust you to not download a cat video or shutdown your computer while I'm trying to do a file transfer ? Can I trust random people with no technical knowledge ?

@stqism 

> @tux3 I'm speaking to @irungentoo in real life and he explained that even if you added tcp relay support, no one would connect to you in the first place. 

Then this goes from potentially dangerous to mostly useless. I guess with bold red angry blinking warnings it could work. If it mitigate risk of disruption of "service" like @zetok says.

**Comment by stqism** [comment]
@tux3 Wait until @irungentoo makes an actual core feature for forwarding, I've spoken to him about encouraging @zetok in the future for dumb ideas.

**Comment by zetok** [comment]
> > If it's not done in reliable way, then Tox will inevitably go down, whenever someone who doesn't like Tox will feel like taking Tox down.
> 
> I agree, but that's not reason enough to add that option.

Yes, I was just pointing out that it can't be really used as reason against.

> > > That would be a solution, but unless you plan to run a Qt tox client on a server, that still implies having a relay on a home computer behind a home connection.
> > 
> > And what could be possibly wrong with that?
> 
> People might decide to use their home connection do download 350 TB of cat pictures or maybe shutdown their computers so it doesn't get too hot with qTox running. A server is supposed to be more or less consistently available.

That's why there should be a big red warning, some words of explanation end perhaps link to page with more info/explanation with colourful pictures.

> > Currently public Tox bootstraps (which have TCP-server enabled) generate ~250kbps of traffic.. That's really a lot, considering that they're public, and like every client uses them to bootstrap. And that CPU usage! I'm sure it goes into cosmic values…
> 
> Yes, the network is still small.

And I forgot to add that as far as I can see, average speed of internet connections increases as time flows, so..

> > I'm not sure about your internet connection, but I myself cold have that on my home network without problems. And my internet connection isn't even that good..
> 
> But can I trust you to not download a cat video or shutdown your computer while I'm trying to do a file transfer ?

Nope. You, or anyone else cannot have a trust. That's why it should be build reliably, and there should be warning + explanation of expectations ("requirements" ?).

On the other side though, I don't really see a point in making it simply hard for people to contribute in some meaningful way to Tox network.

Tox will grow, thus there will be constantly increasing demand for available TCP relays. It's a matter of whether really big demand by millions of users could be satisfied by relatively small group of people who know how to operate in *nix environment, have means to setup server that will be available, and who actually will do it.

On the contrary, there's quite big number of people who aren't really that technical, but could greatly help, by running **easy to use** piece of software that would have TCP-server functionality ready to enable, in easy way, provided that user would make an **informed decision** whether something like that should be ran on ones hardware.

**Comment by stqism** [comment]
... 

@irungentoo still has to add the ability to figure out the best connections and only use them so actual clients can route around people attempting to be tcp relays in desktop clients, stop trying to waste @tux3's time with useless issues.

**Comment by zetok** [comment]
It's not useless, it already helped to give a birth to idea of relying traffic for friends.

Possibly more in future, but since apparently core lacks features, it should be there as a reminder that something like that is needed.

**Comment by tux3** [comment]
@stqism 

> I've spoken to him about encouraging @zetok in the future for dumb ideas. 

That's not exactly helping, stqism.

@zetok 

> Yes, I was just pointing out that it can't be really used as reason against.

Alright.

> That's why there should be a big red warning, some words of explanation end perhaps link to page with more info/explanation with colourful pictures.

Nobody reads pages with more info, as far as I know. Not the persons I'm concerned about, at least. But it's always helpful to have, of course.

> Nope. You, or anyone else cannot have a trust. That's why it should be build reliably, and there should be warning + explanation of expectations ("requirements" ?).

True.

> On the other side though, I don't really see a point in making it simply hard for people to contribute in some meaningful way to Tox network.

If something has too much potential for misuse, then perhaps not keeping it in the spotlight would be wiser. The road to hell is paved with good intentions. But if it doesn't actually harm the network, then that point is moot.

> Tox will grow, thus there will be constantly increasing demand for available TCP relays. It's a matter of whether really big demand by millions of users could be satisfied by relatively small group of people who know how to operate in *nix environment, have means to setup server that will be available, and who actually will do it.

If I take the Tor network as a point of comparison, they do have problems with the many slow relays. People with unstable connections or <200kb/s are advised not to run a relay afaik.

> On the contrary, there's quite big number of people who aren't really that technical, but could greatly help, by running easy to use piece of software that would have TCP-server functionality ready to enable, in easy way, provided that user would make an informed decision whether something like that should be ran on ones hardware.

Again, see the Tor network. The most helpful users are actually technically knowledgeable people running relays on real servers. The small home relays are mostly just slowing down the network and bloating the consensus.

@stqism 

> stop trying to waste @tux3's time with useless issues. 

It's quite alright, thank you very much. I'll gladly take talking over insults any day.

**Comment by stqism** [comment]
@zetok He already had the idea, but being one guy, he can't do everything at once. 

I think it's great you want to help Tox, but bugging down core and client developers when something makes zero sense is not the way to do it.

One thing that would be greatly appreciated would be pull requests to respective repos themselves, most of our developers are bogged down with lots of features and fixes to implement. 

If you'd like to help with toxcore, the repo is https://github.com/irungentoo/toxcore. 

Happy toxing!

**Comment by tux3** [comment]
> I think it's great you want to help Tox, but bugging down core and client developers when something makes zero sense is not the way to do it.

And shooing people off when they try to help is not the way either.

> One thing that would be greatly appreciated would be pull requests to respective repos themselves, most of our developers are bogged down with lots of features and fixes to implement. 

Indeed. The <a href="https://github.com/tux3/qTox/graphs/contributors">contributors</a> page is very helpful in pointing out who actually contributes to this repo.

Happy toxing!

**Comment by stqism** [comment]
@tux3 Showed @irungentoo the thread IRL, he said" he misunderstood what I said", so I took it on myself to convoy what @irungentoo himself told me. 

@zetok decided to argue.

**Comment by zetok** [comment]
Thanks @stqism for pointing toxcore repo, I've already contributed to it today by making an issue there.

Edit: If I misunderstood, I really would like to hear proper explanation if possible.
